lms/djangoapps/course_api/tests/test_views.py

```diff
@@ -348,3 +356,45 @@ def test_list_all_with_search_term(self):
         self.assertIn('results', res.data)
         self.assertNotEqual(res.data['results'], [])
         self.assertEqual(res.data['pagination']['count'], 1)  # Should list a single course
+
+    def test_too_many_courses(self):
+        """
+        Test that search results are limited to 100 courses, and that they don't
+        blow up the database.
+        """
+
+        ContentTypeGatingConfig.objects.create(
+            enabled=True,
+            enabled_as_of=datetime(2018, 1, 1),
+        )
+
+        CourseDurationLimitConfig.objects.create(
+            enabled=True,
+            enabled_as_of=datetime(2018, 1, 1),
+        )
+
+        course_ids = []
+
+        # Create 300 courses across 30 organizations
+        for org_num in range(10):
+            org_id = 'org{}'.format(org_num)
+            for course_num in range(30):
+                course_name = 'course{}.{}'.format(org_num, course_num)
+                course_run_name = 'run{}.{}'.format(org_num, course_num)
+                course = CourseFactory.create(org=org_id, number=course_name, run=course_run_name, emit_signals=True)
+                course_ids.append(course.id)
+
+        self.setup_user(self.audit_user)
+
+        # These query counts were found empirically
+        query_counts = [174, 196, 226, 335, 286, 316, 504, 571, 639, 723, 673]
+        ordered_course_ids = sorted([str(cid) for cid in (course_ids + [c.id for c in self.courses])])
+
+        for page in range(1, 12):
+            with self.assertNumQueries(query_counts[page-1]):
```

Comment:
You seem to have plenty of reviewers, so I'm just skimming this PR. One small nit: These query counts are going to need to get adjusted the every time someone tweaks any piece of middleware. If there's an easy way to push access check query counting down a layer to a place where it's more isolated from shared authn/middleware/etc. queries, that would be nice.

Reply:
I ended up in the vicinity of 50 queries per page, ~30 of which are from the test code. I'm going to leave the breakdown for another time.